tests/no_parallel.rs

```diff
@@ -0,0 +1,27 @@
+#![cfg(not(feature = "parallel"))]
+
+use std::rc::Rc;
+
+use specs::{storage::VecStorage, Builder, Component, Join, World, WorldExt};
+
+#[derive(PartialEq)]
+struct CompNonSend(Rc<u32>);
+
+impl Component for CompNonSend {
+    type Storage = VecStorage<Self>;
+}
+
+#[test]
+fn non_send_component_is_accepted() {
+    let mut world = World::new();
+    world.register::<CompNonSend>();
+
+    let entity = world
+        .create_entity()
+        .with(CompNonSend(Rc::new(123)))
```

Comment:
@chinedufn I've made this test different to the one in `shred` -- in `shred` it tests resources, but in `specs` we have to test components.

This part doesn't compile yet -- even if the backing `UnprotectedStorage` doesn't have the `Send + Sync` requirement, the `Component` itself needs to have that relaxed. But when relaxing it in the `Builder` trait (which it is done above), the constraint ripples to other types as well (many, many ripples).

Since there's many ripples, and it's not maintainable to create two versions of the same code (one with and one without `Send + Sync`), I kind of want to wait for [trait aliases](https://github.com/rust-lang/rust/issues/41517) to be stabilized so that we can do something like:

```rust
#![cfg(feature = "parallel")]
trait ComponentBound = Component + Send + Sync;
#![cfg(not(feature = "parallel"))]
trait ComponentBound = Component;
```

and use `ComponentBound` where it's needed.

Reply:
Wait a minute, do we need to support the `Component` being `!Send + !Sync`? or is it enough to support the storage not being `Send + Sync`?

Reply:
Okay, the experiment shows it's possible -- the test passes. But the duplication is still unwieldly.
Shall attempt a less duplicated approach in a bit.

Reply:
e1477cf removes the duplication, but forces closures to explicitly specify the type of the `world` variable: 1ed2912, which is an ergonomic breakage I would like to avoid.

Reply:
okay, so b4e83ce uses a macro to generate the feature-specific methods (with/without `Send + Sync`). Gonna do a quick summary for reviewers